src/TestFramework/Microsoft.Rest.ClientRuntime.Azure.TestFramework/TestEnvironment.cs

```diff
@@ -322,29 +348,70 @@ private void InteractiveLogin(string tenant, ActiveDirectoryServiceSettings aadS
+            if (!(string.IsNullOrEmpty(this.SubscriptionId)) && !(this.SubscriptionId.Equals("None", StringComparison.OrdinalIgnoreCase)))
+            {   
+                if (subscriptionList.Any<SubscriptionInfo>())
                 {
-                    StringBuilder sb = new StringBuilder("List of subscriptions retrieved:\r\n");
-                    foreach (SubscriptionInfo subInfo in subscriptionList)
+                    var matchedSubs = subscriptionList.Where((sub) => sub.SubscriptionId.Equals(this.SubscriptionId, StringComparison.OrdinalIgnoreCase));
+                    if (matchedSubs.IsAny<SubscriptionInfo>())
+                    {
+                        matchedSubscriptionId = matchedSubs.FirstOrDefault().SubscriptionId;
+                    }
+                    else
                     {
-                        sb.AppendLine(subInfo.SubscriptionId);
+                        foreach (SubscriptionInfo subInfo in subscriptionList)
+                        {
+                            subs += subInfo.SubscriptionId + ",";
+                        }
                     }
+                }
 
-                    //TODO: Find a way to get userId that is being used to retrive subscriptions
-                    //Currently if there occurs a situation, the subscription list retrieved does not match the supplied subscriptionIds
-                    //There is not way to let the user know as to for what user the subscription list was retrieved.
-                    string exceptionString = string.Format("Either no subscription was provided in connection string (e.g. TEST_CSM_ORGID_AUTHENTICATION=SubscriptionId=<subscritionId> or " +
-                        "The provided SubscriptionId in connections string - '{0}' does not match the list of subscriptions associated with the account \r\n {1}", this.SubscriptionId, sb.ToString());
+                if (string.IsNullOrEmpty(matchedSubscriptionId))
+                {
+                    sb.AppendLine(string.Format("SubscriptionList:'{0}' retrieved for the user/spn id '{1}', do not match with the provided subscriptionId '{2}' in connection string",
+                        subs, callerId, this.SubscriptionId));
 
-                    throw new Exception(exceptionString);
+                    throw new Exception(sb.ToString());
+                }
+            }
+            else
+            {
+                // The idea is in case if no match was found, we check if subscription Id was provided in connection string, if not
+                // we then check if the retrieved subscription list has exactly 1 subscription, if yes we will just use that one.
+                if (string.IsNullOrEmpty(this.SubscriptionId))
+                {
+                    if (subscriptionList.Count() == 1)
+                    {
+                       this.ConnectionString.KeyValuePairs[ConnectionStringKeys.SubscriptionIdKey] = subscriptionList.First<SubscriptionInfo>().SubscriptionId;
```

Comment:
Please file an issue to add a test for this scenario.  We should be able to mock the APIs to make this possible to test.